cardano-cli/src/Cardano/CLI/Shelley/Run/Genesis/Types.hs

```diff
@@ -0,0 +1,185 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MonoLocalBinds #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+{- HLINT ignore "Monoid law, left identity" -}
+
+module Cardano.CLI.Shelley.Run.Genesis.Types
+  ( OutputShelleyGenesis(..)
+  , OutputShelleyGenesisStaking(..)
+  , ListMap(..)
+  , toOutputTemplate
+  ) where
+
+import Cardano.CLI.Shelley.Run.Genesis.ListMap (ListMap(..))
+import Cardano.CLI.Shelley.Run.Genesis.LazyToJson (LazyToJson(..))
+import Cardano.Ledger.Address (Addr)
+import Cardano.Ledger.BaseTypes (PositiveUnitInterval, Network)
+import Cardano.Ledger.Coin (Coin)
+import Cardano.Ledger.Era (Era(Crypto))
+import Cardano.Ledger.Keys (KeyHash, KeyRole(Genesis), GenDelegPair)
+import Cardano.Ledger.Shelley.PParams ( PParams )
+import Cardano.Slotting.Slot (EpochSize(..))
+import Data.Aeson (ToJSON(..), (.=))
+import Data.Eq (Eq)
+import Data.Function (($))
+import Data.Monoid (Monoid(..), mconcat)
+import Data.Semigroup ((<>))
+import Data.Time (NominalDiffTime, UTCTime(..))
+import Data.Word (Word32, Word64)
+import GHC.Generics (Generic)
+import Text.Show (Show)
+
+import qualified Cardano.Ledger.Crypto as Ledger
+import qualified Cardano.Ledger.Keys as Ledger
+import qualified Cardano.Ledger.Shelley.Genesis as Ledger
+import qualified Cardano.Ledger.Shelley.TxBody as Ledger
+import qualified Data.Aeson as Aeson
+import qualified Data.Map as Map
+
+data OutputShelleyGenesis era = OutputShelleyGenesis
+  { sgSystemStart :: !UTCTime
+  , sgNetworkMagic :: !Word32
+  , sgNetworkId :: !Network
+  , sgActiveSlotsCoeff :: !PositiveUnitInterval
+  , sgSecurityParam :: !Word64
+  , sgEpochLength :: !EpochSize
+  , sgSlotsPerKESPeriod :: !Word64
+  , sgMaxKESEvolutions :: !Word64
+  , sgSlotLength :: !NominalDiffTime
+  , sgUpdateQuorum :: !Word64
+  , sgMaxLovelaceSupply :: !Word64
+  , sgProtocolParams :: !(PParams era)
+  , sgGenDelegs :: !(ListMap (KeyHash 'Genesis (Crypto era)) (GenDelegPair (Crypto era)))
```

Comment:
The real solution would be to modify the `ShelleyGenesis` datatype in the ledger i.e genDelegs and initialFunds. Changing them from `Map`s to `[(*,*)]`. This should not be a problem, can you talk to @JaredCorduan about doing this? We can then avoid defining the `ListMap` type and this much (if not all) of this module.

Reply:
Are you proposing to change the format of the JSON file to use `[["key1", "value1"], ["key2", "value2]]` instead of `{"field1": "value1", "field2", "value2"}`?

Reply:
My guess would be that `sgGenDelegs` does not need to be a lazy `ListMap`, since that field always contains a very small `Map`. The other two fields below however can be safely switched to `ListMap`, since ledger doesn't even use them.

Reply:
Using whatever is defined in `cardano-ledger` now.